pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h

```diff
@@ -178,6 +183,48 @@ triggerMovementToHoldPosition()
   JointTrajectoryController::setHoldPosition(this->time_data_.readFromRT()->uptime);
 }
 
+template <class SegmentImpl, class HardwareInterface>
+inline bool PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+checkStates(const ros::Duration& period) const
+{
+  return (cartesian_speed_monitor_->cartesianSpeedIsBelowLimit(JointTrajectoryController::old_desired_state_.position,
+                                                               JointTrajectoryController::desired_state_.position,
+                                                               period.toSec(),
+                                                               0.25 /*limit */));  // TODO: HSL
+}
+
+template <class SegmentImpl, class HardwareInterface>
+void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+reactToFailedStateCheck(const ros::Time& curr_uptime)
+{
+  triggerCancellingOfActiveGoal();
+
+  stop_traj_builder_->buildTrajectory(JointTrajectoryController::old_time_data_.uptime.toSec(),
+                                      JointTrajectoryController::old_desired_state_,
+                                      stop_traj_velocity_violation_.get(), nullptr);
+
+  JointTrajectoryController::updateStates(curr_uptime, stop_traj_velocity_violation_.get());
+
+  JointTrajectoryController::curr_trajectory_box_.set(stop_traj_velocity_violation_);
+}
+
+template <class SegmentImpl, class HardwareInterface>
+inline void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+triggerCancellingOfActiveGoal()
+{
+  RealtimeGoalHandlePtr active_goal(JointTrajectoryController::rt_active_goal_);
+  if (!active_goal)
+  {
+    return;
+  }
+  JointTrajectoryController::rt_active_goal_.reset();
+  active_goal->gh_.setCanceled();
+  // TODO: Instead of the line above, I actually want to do this:
+  //      active_goal->preallocated_result_->error_code = control_msgs::FollowJointTrajectoryResult::PATH_TOLERANCE_VIOLATED;
+  //      active_goal->setCanceled(active_goal->preallocated_result_);
+  // Unfortunately this does not work because sometimes the goal does not seems to get cancelled.
+  // It has to be investigated why! -> Probably a threading problem.
```

Comment:
Das Problem könntest du umgehen, indem du den RealTimeGoalHandlePtr zurückgibst und an die hold-Trajektorie weitergibst.
```
template <class SegmentImpl, class HardwareInterface>
void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
reactToFailedStateCheck(const ros::Time& curr_uptime)
{
  RealtimeGoalHandlePtr gh = triggerCancellingOfActiveGoal();

  stop_traj_builder_->buildTrajectory(JointTrajectoryController::old_time_data_.uptime.toSec(),
                                      JointTrajectoryController::old_desired_state_,
                                      stop_traj_velocity_violation_.get(), gh);
```

Reply:
Natürlich ist das Problem damit nicht gelöst. Es besteht genauso in der update()-Funktion bei den tolerance-checks:
```
          rt_segment_goal->setAborted(rt_segment_goal->preallocated_result_);
          rt_active_goal_.reset();
```
Auch hier keine Garantie, dass das durchgeht, bevor `rt_segment_goal` out of scope geht.

Reply:
I suggest to close this because it is not a problem caused by us and should be handled by a different PR. I added a comment above the corresponding code lines. Is this enough for you?

Reply:
Yes. You could reference https://github.com/ros-controls/ros_controllers/issues/174

Reply:
> Yes. You could reference [ros-controls/ros_controllers#174](https://github.com/ros-controls/ros_controllers/issues/174)

What do you mean? Should I reference the PR in a comment in code?

Reply:
Added. -> Done.